File: app/src/main/java/com/arcuscomputing/db/FavouritesDbHelper.java
package com.arcuscomputing.db;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.database.sqlite.SQLiteStatement;

import com.arcuscomputing.models.WordModel;

import java.util.LinkedList;
import java.util.List;

public class FavouritesDbHelper extends SQLiteOpenHelper {

    public final static String OPTION_SORT_ALPHA_ASC = "word  ASC";
    public final static String OPTION_SORT_ALPHA_DESC = "word DESC";
    public final static String OPTION_SORT_DATE_ASC = "date_added ASC";
    public final static String OPTION_SORT_DATE_DESC = "date_added DESC";
    private static final int DATABASE_VERSION = 3;
    private static final String DICTIONARY_TABLE_CREATE =
            "CREATE TABLE favourites (word TEXT, definition TEXT, date_added DATETIME default CURRENT_TIMESTAMP)";

    private static final String INSERT_SQL =
            "INSERT INTO favourites (word, definition) values (?,?)";

    private static final String FAVOURITE_QUERY =
            "SELECT * FROM favourites WHERE word = ? AND definition = ?";

    //TODO paramaterise?
    private static final String ALL_FAVOURITES_QUERY =
            "SELECT * FROM favourites ORDER BY ";

    private static final String DELETE_FAVOURITE_QUERY =
            "DELETE FROM favourites WHERE word = ? AND definition = ?";

    private static final String DELETE_ALL_FAVOURITES_QUERY = "DELETE FROM favourites";

    private SQLiteDatabase db;

    private SQLiteStatement insertStatement;

    public FavouritesDbHelper(Context context) {
        super(
                context,
                context.getPackageName().contains("pro")
                        ? "arcusdictionarypro.db"
                        : "arcusdictionary.db",
                null,
                DATABASE_VERSION);

        synchronized (this) {
            this.db = this.getWritableDatabase();
        }

        this.insertStatement = this.db.compileStatement(INSERT_SQL);
    }

    @Override
    public synchronized void onCreate(SQLiteDatabase db) {
        db.execSQL(DICTIONARY_TABLE_CREATE);
    }

    @Override
    public synchronized void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        db.execSQL("DROP TABLE IF EXISTS favourites");
        onCreate(db);
    }

    public synchronized void insertfavourite(String word, String definition) {

        this.insertStatement.bindString(1, word);
        this.insertStatement.bindString(2, definition);
        this.insertStatement.execute();
    }

    public synchronized boolean isFavourite(String word, String definition) {

        Cursor result = db.rawQuery(FAVOURITE_QUERY, getArgsFromWordAndDef(word, definition));

        int count = result.getCount();
        result.close();

        return count == 1;
    }

    private String[] getArgsFromWordAndDef(String word, String definition) {

        String[] args = new String[2];
        args[0] = word;
        args[1] = definition;

        return args;
    }

    public synchronized boolean deleteFromFavourites(String word, String definition) {
        Cursor cursor = db.rawQuery(DELETE_FAVOURITE_QUERY, getArgsFromWordAndDef(word, definition));
        int count = cursor.getCount();
        cursor.close();


        return count == 1;
    }

    public synchronized List<WordModel> getAllFavourites(String sortMethod) {
        Cursor cursor = db.rawQuery(ALL_FAVOURITES_QUERY + sortMethod, null);
        List<WordModel> results = new LinkedList<WordModel>();

        while (cursor.moveToNext()) {
            WordModel wm = new WordModel();
            wm.setWord(cursor.getString(cursor.getColumnIndex("word")));
            String def = cursor.getString(cursor.getColumnIndex("definition"));
            wm.setDefinition(def);
            results.add(wm);
        }

        cursor.close();
        return results;
    }

    public synchronized void deleteAllFavourites() {
        Cursor cursor = db.rawQuery(DELETE_ALL_FAVOURITES_QUERY, null);
        int count = cursor.getCount();
        cursor.close();
    }

}	

File: app/src/main/java/com/arcuscomputing/FavouritesDbHelper.kt
package com.arcuscomputing

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.database.sqlite.SQLiteStatement

class FavouritesDbHelper(context: Context) : SQLiteOpenHelper(
    context,
    if (context.packageName.contains("pro")) "arcusdictionarypro.db" else "arcusdictionary.db",
    null,
    DATABASE_VERSION
) {

    enum class SortOrder(internal val sql: String) {
        ALPHA_ASC("word ASC"),
        ALPHA_DESC("word DESC"),
        DATE_ASC("date_added ASC"),
        DATE_DESC("date_added DESC"),
    }

    private val db: SQLiteDatabase
    private val insertStatement: SQLiteStatement

    init {
        db = writableDatabase
        insertStatement = db.compileStatement(INSERT_SQL)
    }

    override fun onCreate(db: SQLiteDatabase) {
        db.execSQL(DICTIONARY_TABLE_CREATE)
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        db.execSQL("DROP TABLE IF EXISTS favourites")
        onCreate(db)
    }

    @Synchronized
    fun insertFavourite(word: String, definition: String) {
        insertStatement.bindString(1, word)
        insertStatement.bindString(2, definition)
        insertStatement.execute()
    }

    @Synchronized
    fun isFavourite(word: String, definition: String): Boolean {
        val cursor = db.rawQuery(COUNT_QUERY, arrayOf(word, definition))
        val count = if (cursor.moveToFirst()) cursor.getInt(0) else 0
        cursor.close()
        return count > 0
    }

    @Synchronized
    fun deleteFromFavourites(word: String, definition: String): Boolean {
        val rowsAffected = db.delete("favourites", "word = ? AND definition = ?", arrayOf(word, definition))
        return rowsAffected == 1
    }

    @Synchronized
    fun getAllFavourites(sortOrder: SortOrder): List<WordModel> {
        val cursor = db.rawQuery("SELECT * FROM favourites ORDER BY ${sortOrder.sql}", null)
        val results = mutableListOf<WordModel>()
        while (cursor.moveToNext()) {
            results.add(WordModel(
                word = cursor.getString(cursor.getColumnIndexOrThrow("word")),
                definition = cursor.getString(cursor.getColumnIndexOrThrow("definition"))
            ))
        }
        cursor.close()
        return results
    }

    @Synchronized
    fun deleteAllFavourites() {
        db.execSQL("DELETE FROM favourites")
    }

    companion object {
        private const val DATABASE_VERSION = 3
        private const val DICTIONARY_TABLE_CREATE =
            "CREATE TABLE favourites (word TEXT, definition TEXT, date_added DATETIME default CURRENT_TIMESTAMP)"
        private const val INSERT_SQL =
            "INSERT INTO favourites (word, definition) values (?,?)"
        private const val COUNT_QUERY =
            "SELECT COUNT(*) FROM favourites WHERE word = ? AND definition = ?"
    }
}
